### DISCUSSION TITLE
Signing Mac64 apps

### DISCUSSION THREAD

**Discussion by borjafdezgauna** [picoe/Eto#1890] [Q&A] [answered]
Hi, I have spent some time trying to sign my EtoForms app on Mac and haven't been able to do it. I generated my Apple Developer certificate (.p12) and have been trying to use codesign on Mac to sign it, but I get the error message "main executable failed strict validation". Anyone has had success signing a netcore3.1 app on Mac?

Thanks in advance!
Borja

**Comment by cwensley** [comment, marked as answer]
Signing/Notarization doesn't work with single file app.  If you set `<PublishSingleFile>` to false, it should work.  Although, I've only tested this on .NET 5.

.NET 6 should resolve that issue, though it will still not be "fully single file".

Hope this helps!

**Comment by borjafdezgauna** [comment]
I think I need to sign the assemblies one by one in the correct order, but it seems to work fine
Thank you very much!!

**Comment by cwensley** [comment]
passing `--deep` to the codesign command line _should_ sign all of the correct files for you.

BTW, did you know that Eto.Platform.Mac64 nuget package can code sign, notarize, and package your app into a .dmg for you?

It is not documented in the wiki yet, but should work for most cases.  See [here](https://github.com/cwensley/Eto/blob/curtis/gridcolumn-tweaks/src/Eto.Mac/build/Mac.targets#L289-L291) for where it is implemented.

In particular, you need at _least_ these properties set in your csproj, something like this:

```xml
  <PropertyGroup Condition="$([MSBuild]::IsOsPlatform(OSX)) AND $(Configuration) == 'Release'">
    <EnableDmgBuild>True</EnableDmgBuild>
    <EnableCodeSigning>True</EnableCodeSigning>
    <EnableNotarization>True</EnableNotarization>
  </PropertyGroup>
```

For notarization, you need to add the apple account to your keychain like so:

```
xcrun altool --store-password-in-keychain-item "AC_PASSWORD" -u "<NotarizationUserName>" -p <secret_password>
```

You can see how it can be automated in GitHub Actions [here](https://github.com/picoe/Eto/blob/develop/.github/workflows/build.yml#L94-L109)

Hope this helps!

**Reply by borjafdezgauna** [reply]
Thank you so much for your help!!

**Reply by borjafdezgauna** [reply]
Thanks to your indications, I was finally able to build and notarize a dmg installer with my application. I am using an Azure pipeline instead of GitHub actions, I was getting this error message when building the app:

`Failed to parse entitlements: AMFIUnserializeXML: syntax error near line 1`

I suspect is related with the line-endings, but it's weird because I could build it on my Mac and it failed on the Azure MacOs VM. Anyway, doing this in my build pipeline allows me to build the app without errors:

`plutil -convert xml1 /Users/runner/.nuget/packages/eto.platform.mac64/2.5.10/build/CodeSign.entitlements`

Just for reference in case someone encounters the same error...

**Reply by borjafdezgauna** [reply]
And I have a question: how can I change the icon used to build the image. It seems to be using the one on the Eto.Mac/build folder instead of the one in the Mac project folder. What is the right way to change the icon? Thanks!!

**Reply by borjafdezgauna** [reply]
> And I have a question: how can I change the icon used to build the image. It seems to be using the one on the Eto.Mac/build folder instead of the one in the Mac project folder. What is the right way to change the icon? Thanks!!

Ok, I think I figured it out. I had it working copying manually the icon file to the output resource folder. Thanks anyway!

**Reply by cwensley** [reply]
@borjafdezgauna hm, not sure why CodeSign.entitlements have incorrect line endings there in the nupkg..  I'll look into it.

As for the Icon, you need to specify it as Content or BundleResource in your csproj.  e.g.

```
<ItemGroup>
  <Content Include="Icon.icns" />
</ItemGroup>
```

**Reply by borjafdezgauna** [reply]
Argh! I thought I had already tested that. It works fine... my bad. Sorry and thank you so much!!
